scipy/sparse/linalg/_interface.py

```diff
@@ -175,11 +175,29 @@ def __init__(self, dtype, shape):
         self.shape = shape
 
     def _init_dtype(self):
-        """Called from subclasses at the end of the __init__ routine.
+        """Determine the dtype by executing `matvec` on an `int8` test vector.
+
+        In `np.promote_types` hierarchy, the type `int8` is the smallest,
+        so we call `matvec` on `int8` and use the promoted dtype of the output
+        to set the default `dtype` of the `LinearOperator`.
+        We assume that `matmat`, `rmatvec`, and `rmatmat` would result in
+        the same dtype of the output given an `int8` input as `matvec`.
+
+        Called from subclasses at the end of the __init__ routine.
         """
         if self.dtype is None:
-            v = np.zeros(self.shape[-1])
-            self.dtype = np.asarray(self.matvec(v)).dtype
+            v = np.zeros(self.shape[-1], dtype=np.int8)
+            try:
+                matvec_v = np.asarray(self.matvec(v))
+            except OverflowError:
+                # Python large `int` promoted to `np.int64`or `np.int32`
+                try:
+                    test_int64 = np.int64(1)  # noqa: F841
+                    self.dtype = np.int64
+                except AttributeError:
+                    self.dtype = np.int32
```

Comment:
I don't know if this will work because a python int can still overflow a long int?
```
In [2]: 10**50*np.array(1, dtype=np.int64)
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In[2], line 1
----> 1 10**50*np.array(1, dtype=np.int64)

OverflowError: Python int too large to convert to C long
```

Reply:
Thanks, great point! 

Maybe I should just give up and turn Python int outside of int8 into a float64? It would also make the code backward compatible. What do you think?

Reply:
testing turn Python int outside of int8 into a float64

Reply:
@j-bowhay casting Python int outside of int8 into a float64 is not always working, so I am back to my original code. It does not cast all the cases accurately, as you point out, but it is still an improvement over the main that is always `np.float64`. I have added a comment to the docs to underscore the possible inaccuracy.